types/src/proof/definition.rs

```diff
@@ -303,6 +304,62 @@ impl AccumulatorConsistencyProof {
     pub fn siblings(&self) -> &[HashValue] {
         &self.siblings
     }
+
+    /// Verifies the proof. The verification consists of two steps. First, the frozen subtree roots
+    /// should match `old_root_hash` and `old_num_leaves`. Second, after combining the frozen
+    /// subtree roots with the siblings we should be able to obtain `new_root_hash`.
+    pub fn verify<H: CryptoHasher>(
+        &self,
+        old_root_hash: HashValue,
+        old_num_leaves: u64,
+        new_root_hash: HashValue,
+        new_num_leaves: u64,
+    ) -> Result<()> {
+        ensure!(
+            old_num_leaves <= new_num_leaves,
+            "Old accumulator ({} leaves) is bigger than new accumulator ({} leaves).",
+            old_num_leaves,
+            new_num_leaves,
+        );
+
+        let old_accumulator =
+            Accumulator::<H>::new(self.frozen_subtree_roots.clone(), old_num_leaves)?;
```

Comment:
It feels weird for the client to pass in old_num_leaves.  Shouldn't the proof know the N in the N << M that it is proving?  Maybe we should store that with the proof itself instead of passing it in?  It's just one more parameter for clients to get wrong.

Reply:
I imagine the way this works is that the client will initially have a waypoint that is a tuple `(version, root_hash)` (basically `LedgerInfo`). Every time the client calls `update_to_latest_ledger`, it receives the same tuple at a newer version, verifies the proof, saves the new `(version, root_hash)` tuple, and discards the old one. So the client always persists the latest tuple. Since the client already knows `N`, putting `N` in the proof might cause some confusion.

So the client would pass in both the old tuple and the new one. Maybe taking two `LedgerInfo`s is better?

Reply:
I don't think it's weird since our merkle proof is similar. A proof can be both existent proof of a specific key and a non-existent proof of many other keys.